### PR TITLE
Various CentOS 7 fixes and some features/changes

### PR DESCRIPTION
**Pull request by foxxx0**
This PR is the result of quite intensive debugging and fixing CentOS 7 compatibility.

It consists of:
 - optionally override GPT `first-lba` header
 - disable `metadata_csum` ext4 feature for CentOS 7 images (not supported by their old `e2fsprogs` version)
 - enable `64bit` ext4 feature on supported architectures
 - switch to `grub2-efi` for efi-bootable CentOS 7 images (their `bootctl`/`systemd-boot` is broken / too old)
 - run the postinstall script with nspawn blockdev access for bootable images

I have tried to group the changes into individual logical commits as best as possible, so reviewing should probably done commit-by-commit.

If you want, I can also split this into multiple PRs, just say so.

Any feedback is welcome and I'm happy to adjust it to meet your requirements.

### CODE REVIEW

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed? please check again if I understood your suggestion correctly.

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi
Comment (foxxx0) [reply mkosi]: fixed

File: mkosi.md
Comment (foxxx0) [reply mkosi.md:245]: done

**Comment by foxxx0** [comment]
Thanks for the detailed review with very constructive feedback.
I have dropped the `epel_release` -> `release_major` rename commit and adjusted the other commits and tried to implement/fix all suggestions.

> > I've never tried to use unified kernel images with grub.
> 
> I don't think it actually supports them. I don't think we need grub-efi to work with unified kernel images either. We just fail early and tell users to use `--without-unified-kernel-images`.

For this I have followed the early call to `die()`/`warn()` within `load_args()` for Clear Linux, i.e. for the `CentOS7-bootable-uefi` config it will print a warning and set `args.with_unified_kernel_images` to `False`. I hope this is what you had in mind.

Please give it another look and let me know if there are still changes required.

Also please advise if you want to have the generic "Bootloader Choice" (grub/systemd-boot) feature implemented within this PR or if that can happen seperatly once this PR is merged.

**Comment by foxxx0** [comment]
rebased onto current master

File: mkosi/__init__.py
Comment (DaanDeMeyer) [inline mkosi/__init__.py:4523]: I'd prefer simply failing early if `--without-unified-kernel-images` is not specified for either the bootloader option or centos 7 (depending on if you're up for adding the bootloader option).

I know I added the Clear Linux automatic fallback but I already removed that (or its in one of the open PRs) that. It adds to much magic and I'd prefer users specifically acknowledging and being okay with centos 7 not doing unified kernel images rather than automatically assume they will be ok with it just because centos 7 happens to not support them.

File: mkosi/__init__.py
Comment (DaanDeMeyer) [inline mkosi/__init__.py:2828]: Do we need this directory specifically? What will grub use as the default?

File: mkosi/__init__.py
Comment (DaanDeMeyer) [inline mkosi/__init__.py:2834]: How does grub decide between BIOS and UEFI? Does it just look at the directory it's being installed in?

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Looks great! A few more comments

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py:2100]: AFAIK the udev files are directly packaged within the `systemd` package and not split out, will check tomorrow.

Comment (DaanDeMeyer) [reply mkosi/__init__.py:2100]: Nvm, if udev is packaged it should be fine. Try to make sure you can run the built image with `mkosi qemu`. If that works then all is good.

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py:2834]: As stated in the comment, it actually does a check for a directory existence of `/sys/firmware/efi`, which fails at least on my build host due to the underlying OS not being booted via UEFI. This in turn causes `grub2-mkconfig` to fall back into `16bit` mode, generating the `linux16 ...` and `initrd16 ...` bootconfig lines.
That is the reason why we need to patch it afterwards.
I've searched a bit through the various manpages and `--help` outputs for the `grub2` utilities but haven't found anything that would allow explicitly forcing an install method while ignoring/overriding the internal checks. IIRC they are in `/etc/grub.d/10-linux`.

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py:2860]: Not sure when I'll have spare time to implement this. For now I'd prefer to get the CentOS 7 fixes merged and maybe open an issue tracking that bootloader choice feature.
I might give it a try sometime in the future since I'll be maintaining our mkosi build pipeline, but no promises.

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py:2828]: > Looks great! A few more comments

Mh, I'm unable to comment on the `mkosi.md` remarks. Indeed I seem to have forgotten to extend that, only reworded the commit message. Gonna fix that tomorrow.

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py:4523]: Fine by me, I'd actually prefer that too.

File: mkosi/__init__.py
Comment (DaanDeMeyer) [reply mkosi/__init__.py:2860]: Could you add the option and only implement it for centos 7? Early failure is fine on all the other distros. The idea behind that is that I want to keep the default behaviour for as much features as possible consistent across distros. Right now, uefi means systemd-boot for everything (except clear but that's a special one). I want to keep it that way ideally. Adding the bootloader option allows us to fail early for centos 7 if --bootloader grub is not specified so users have to add --bootloader grub explicitly for it to work.

Comment (DaanDeMeyer) [reply mkosi/__init__.py:2860]: Also, I'm really interested in what you're using mkosi for. Can you shed any light on that?

File: mkosi/__init__.py
Comment (DaanDeMeyer) [reply mkosi/__init__.py:2828]: Seems you replied to the wrong comment here (or Github did something weird)

File: mkosi/__init__.py
Comment (behrmann) [reply mkosi/__init__.py:888]: > Maybe do `args: List[str] = []` instead. 

This is something that should generally be avoided, because Python initialises default arguments only once, i.e. if the list were mutated inside the function, these would persist function invocations. Only primitive types make good defaults. The idiomatic way to circumvent this is assigning `None` as the default and pushing initialisation into the function body.

**Comment by DaanDeMeyer** [comment]
@foxxx0 Any idea when you'll be able to take another look at this? Ideally we'd get this in before the next release

**Comment by foxxx0** [comment]
@DaanDeMeyer alright so I gave it a quick rework to rebase it onto the current master, which was a bit trickier than expected since you've been quite busy. :stuck_out_tongue: 
I've seem to have resolved the conflicts, but it sure does need another proper review to check if I derped.

I have changed the warning to an early `die()` for centos7 bootable builds in case `--without-unified-kernel-images` was not specified. (along with an instructional message)

I have **not** reworked the `mkfs` part **yet**.

Also I won't be able to implement the bootloader selection stuff anytime soon.
I'll try to get the `mkfs` part done tomorrow but for the other things we either create an issue to track it (feel free to assign me to it) or we keep this PR open until I find some time.
I would highly suggest avoiding the latter, since rebasing will just get trickier and tricker and I feel like the current state should already provide viable fixes to the project best suited for inclusion before the next release.

File: mkosi.md
Comment (behrmann) [reply mkosi.md:245]: This is still missing the example use case and what LBA means (I just looked it up again...)

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:934]:
```suggestion
        if args.architecture in ("x86_64", "aarch64"):
```

Comment (behrmann) [reply mkosi/__init__.py:934]: Also, isn't this set automatically when the image is built on a 64bit system?

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:2094]:
```suggestion
        return int(epel_release) <= 7
```

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:2968]: This line was broken above and should be broken here as well.

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:4044]:
```suggestion
    group.add_argument("--gpt-first-lba", type=int, help='Set the first LBA within GPT Header', metavar='FIRSTLBA')
```
This is a bit shorted and more closely matches the fdisk option

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:4643]:
```suggestion
            die(f"Sorry, CentOS {epel_release} does not support unified kernel images. "
                "You must use --without-unified-kernel-images.")
```
The strings will be automatically concatenated.

**Review by behrmann — COMMENTED** [review]
Quickly looked, but haven't checked the logic yet.

File: mkosi/__init__.py
Comment (DaanDeMeyer) [inline mkosi/__init__.py:2943]: Don't we need to actually install grub as well? I might be missing something obvious but we seem to be missing a grub-install call no?

Comment (foxxx0) [reply mkosi/__init__.py:2943]: The necessary EFI binaries are copied to the ESP upon package installation. So all that's left is to give them a config file to instruct where and how to boot from.

When dealing with BIOS boot, you need to write some magic bytes to the MBR and for GPT >2TB you need to actually have that 1MiB gap within the first 2TB of the disk space so that grub can embed the required code into there because it won't fit into the MBR for this scenario.

I will test and validate that again later today, but so far it has worked on CentOS 7 + EFI + grub2 without a call to `grub2-install`.

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py:934]: Depending on the locally used `e2fsprogs` release it might, but not necessarily (defaults have changed over time). For 64bit target architectures and reproducibility purposes I think it's best to explicityly enable it.

I've committed the suggested change.

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py]: Please check if I understood your remark correctly, I was a bit confused by the wording. :confused:

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py]: Reworked the `mkfs_<filesystem>()` methods to `mkfs_<filesystem>_cmd()` as suggested, please have a look.

File: mkosi.md
Comment (behrmann) [inline mkosi.md:248]: Just a nitpick, but I think this is more in line with the tone of the rest of the docs.
```suggestion
: Override the first usable LBA (Logical Block Address) within the
  GPT header. This defaults to `2048` which is the desired value,
  however, some tools, e.g. the `prl_disk_tool` utility from the
  Parallels virtualization suite require this to be set to `34`, otherwise
  they might fail to resize the disk image and/or partitions inside it.
```
The last two lines are not needed after the excellent example.

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:915]: This logic can be straightened out a bit. If you move the block were `cmdline` is set via `mkfs_*_cmd` before the one handling `mkfs_args` you can add directly to `cmdline` instead of having `mkfs_args` extra.

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:936]:
```suggestion
    run(cmdline + mkfs_args + [dev])
```
or even just `run(cmdline + [dev])` if you take the above suggestion.

File: mkosi/__init__.py
Comment (behrmann) [inline mkosi/__init__.py:2962]: Almost, sorry if my previous wording was unclear. I was referring to line 918.
```suggestion
            elif (args.distribution in (Distribution.centos, Distribution.centos_epel) and 
                  is_older_than_centos8(args.release)):
```
From what I see parentheses are generally preferred in Python over line breaks, sine they explicitly group the predicate.

**Review by behrmann — COMMENTED** [review]
Nice work!

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py]: Great suggestions, thanks. Why didn't I think of that...

File: mkosi/__init__.py
Comment (foxxx0) [reply mkosi/__init__.py]: Ah, that makes more sense indeed.

**Comment by foxxx0** [comment]
Thanks @behrmann for the prompt review!
I've addressed your remarks and force-pushed.

**Review by behrmann — APPROVED** [review]
From my side this is good to go, if the tests pass.

**Comment by DaanDeMeyer** [comment]
I would have liked to add least add the bootloader argument so we don't have to change behavior later to make things consistent but I'm a bit swamped right now so let's add this as is so we can make a new release.

**Pull request merged**
